app/cdap/components/CreateAndEditRecipeForm/index.tsx

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, { FormEvent, useEffect, useState } from 'react';
+import { FormControl } from '@material-ui/core';
+import T from 'i18n-react';
+import {
+  ICreateAndEditRecipeFormProps,
+  IRecipeData,
+} from 'components/CreateAndEditRecipeForm/types';
+import {
+  FormFieldWrapper,
+  NormalLabel,
+  StyledForm,
+  StyledRecipeNameTextField,
+  StyledDescriptionTextArea,
+  StyledCancelButton,
+  StyledSaveButton,
+  StyledFormButtonWrapper,
+  getLabelStyle,
+} from 'components/CreateAndEditRecipeForm/styledComponents';
+import { getCurrentNamespace } from 'services/NamespaceStore';
+import MyDataPrepApi from 'api/dataprep';
+
+export default function({
+  recipeData,
+  setIsCreateAndEditRecipeFormOpen,
+  setSnackbar,
+  recipeFormAction,
+}: ICreateAndEditRecipeFormProps) {
+  const EditRecipeFormAction = 'editRecipe';
+  const [isNameError, setIsNameError] = useState(false);
+  const StyledLabel = getLabelStyle(isNameError);
+  const [recipeFormData, setRecipeFormData] = useState<IRecipeData>({
+    recipeName: '',
+    description: '',
+    directives: [],
+  });
+
+  const [isSaveDisable, setIsSaveDisable] = useState<boolean>(true);
+
+  const recipeSteps = ['uppercase: body1', 'titlecase: body2'];
+
+  useEffect(() => {
+    if (isNameError) {
+      setIsSaveDisable(true);
+    }
+  }, [isNameError]);
+
+  useEffect(() => {
+    setRecipeFormData(recipeData);
+  }, [recipeData]);
+
+  useEffect(() => {
+    if (
+      recipeFormData.recipeName === '' ||
+      recipeFormData.description === '' ||
+      recipeFormData.recipeName?.trim().length === 0 ||
+      recipeFormData.description?.trim().length === 0
+    ) {
+      setIsSaveDisable(true);
+    } else {
+      setIsNameError(false);
+      setIsSaveDisable(false);
+    }
+  }, [recipeFormData]);
```

Comment:
I think this effect can be refactored as a wrapper function around `setRecipeFormData` like the following:

```js
function handleSetRecipeFormData (recipeFormData: IRecipeData): void {
  setRecipeFormData(recipeFormData);
  if (
      recipeFormData.recipeName === '' ||
      recipeFormData.description === '' ||
      recipeFormData.recipeName?.trim().length === 0 ||
      recipeFormData.description?.trim().length === 0
    ) {
      setIsSaveDisable(true);
    } else {
      setIsNameError(false);
      setIsSaveDisable(false);
    }
}
```

Then this `handleSetRecipeFormData` function can be used instead of `setFormData` inside event handlers.

This refactor would reduce the number of times the components needs to re render.

See next comments.

Reply:
used handleSetRecipeFormData instead of useEffect